File: eneter-protobuf-android/src/eneter/protobuf/ProtoBufSerializer.java
package eneter.protobuf;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;

import com.google.protobuf.ByteString;
import com.google.protobuf.GeneratedMessage;
import com.google.protobuf.InvalidProtocolBufferException;

import eneter.messaging.dataprocessing.serializing.ISerializer;
import eneter.messaging.endpoints.typedmessages.VoidMessage;
import eneter.messaging.endpoints.typedmessages.internal.ReliableMessage;
import eneter.messaging.messagingsystems.composites.monitoredmessagingcomposit.*;
import eneter.messaging.nodes.broker.*;
import eneter.messaging.nodes.channelwrapper.WrappedData;
import eneter.protobuf.EneterProtoBufDeclarations.*;

public class ProtoBufSerializer implements ISerializer
{

    @Override
    public <T> Object serialize(T dataToSerialize, Class<T> clazz)
            throws Exception
    {
        Object aSerializedData;
        
        // If it is an internal Eneter type adapt it for ProtoBuf
        if (clazz == WrappedData.class)
        {
            aSerializedData = serializeWrappedData((WrappedData)dataToSerialize);
        }
        else if (clazz == BrokerMessage.class)
        {
            aSerializedData = serializeBrokerMessage((BrokerMessage)dataToSerialize);
        }
        else if (clazz == ReliableMessage.class)
        {
            aSerializedData = serializeReliableMessage((ReliableMessage)dataToSerialize);
        }
        else if (clazz == MonitorChannelMessage.class)
        {
            aSerializedData = serializeMonitorChannelMessage((MonitorChannelMessage)dataToSerialize);
        }
        else if (clazz == VoidMessage.class)
        {
            aSerializedData = serializeVoidMessage((VoidMessage)dataToSerialize);
        }
        else
        {
            // If it is not internal Eneter message use directly ProtoBuf serializer.
            GeneratedMessage aDataToSerialize = (GeneratedMessage)dataToSerialize;
            aSerializedData = aDataToSerialize.toByteArray();
        }
        
        return aSerializedData;
    }

    @Override
    public <T> T deserialize(Object serializedData, Class<T> clazz)
            throws Exception
    {
        T aDeserializedObject;

        // If it is an internal Eneter type, adapt it from the proto type.
        if (clazz == WrappedData.class)
        {
            aDeserializedObject = clazz.cast(deserializeWrappedData((byte[])serializedData));
        }
        else if (clazz == BrokerMessage.class)
        {
            aDeserializedObject = clazz.cast(deserializeBrokerMessage((byte[])serializedData)); 
        }
        else if (clazz == ReliableMessage.class)
        {
            aDeserializedObject = clazz.cast(deserializeReliableMessage((byte[])serializedData)); 
        }
        else if (clazz == MonitorChannelMessage.class)
        {
            aDeserializedObject = clazz.cast(deserializeMonitorChannelMessage((byte[])serializedData)); 
        }
        else if (clazz == VoidMessage.class)
        {
            aDeserializedObject = clazz.cast(deserializeVoidMessage((byte[])serializedData));
        }
        else
        {
            // If it is not an Eneter internal type then use directly protobuf.
            Method aParsingMethod = clazz.getMethod("parseFrom", byte[].class);
            aDeserializedObject = clazz.cast(aParsingMethod.invoke(null, (byte[])serializedData));
        }
        
        return aDeserializedObject;
    }
    
    
    private byte[] serializeWrappedData(WrappedData data)
    {
        WrappedDataProto.Builder aBuilder = WrappedDataProto.newBuilder()
                .setAddedData((String)data.AddedData);
        if (data.OriginalData != null)
        {
            if (data.OriginalData instanceof String)
            {
                aBuilder.setOriginalDataStr((String)data.OriginalData);
            }
            else
            {
                aBuilder.setOriginalDataBin(ByteString.copyFrom((byte[]) data.OriginalData));
            }
        }
        WrappedDataProto aWrappedDataProto = aBuilder.build();

        return aWrappedDataProto.toByteArray();
    }

    private WrappedData deserializeWrappedData(byte[] data) throws InvalidProtocolBufferException
    {
        WrappedDataProto aWrappedDataProto = WrappedDataProto.parseFrom(data);
        WrappedData aWrappedData = new WrappedData();
        aWrappedData.AddedData = aWrappedDataProto.getAddedData();
        if (aWrappedDataProto.hasOriginalDataStr())
        {
            aWrappedData.OriginalData = aWrappedDataProto.getOriginalDataStr();
        }
        else if (aWrappedDataProto.hasOriginalDataBin())
        {
            aWrappedData.OriginalData = aWrappedDataProto.getOriginalDataBin().toByteArray();
        }

        return aWrappedData;
    }


    private byte[] serializeBrokerMessage(BrokerMessage data)
    {
        BrokerMessageProto.Builder aBuilder = BrokerMessageProto.newBuilder()
                .setRequest(data.Request.toString())
                .addAllMessageTypes(Arrays.asList(data.MessageTypes));
        if (data.Message != null)
        {
            if (data.Message instanceof String)
            {
                aBuilder.setMessageStr((String)data.Message);
            }
            else
            {
                aBuilder.setMessageBin(ByteString.copyFrom((byte[])data.Message));
            }
        }
        BrokerMessageProto aBrokerMessageProto = aBuilder.build();
        
        return aBrokerMessageProto.toByteArray();
    }

    private BrokerMessage deserializeBrokerMessage(byte[] data) throws InvalidProtocolBufferException
    {
        BrokerMessageProto aBrokerMessageProto = BrokerMessageProto.parseFrom(data);
        
        List<String> aMessageTypes = aBrokerMessageProto.getMessageTypesList();
        
        BrokerMessage aBrokerMessage = new BrokerMessage();
        aBrokerMessage.Request = EBrokerRequest.valueOf(aBrokerMessageProto.getRequest());
        aBrokerMessage.MessageTypes = aMessageTypes.toArray(new String[aMessageTypes.size()]);
        if (aBrokerMessageProto.hasMessageStr())
        {
            aBrokerMessage.Message = aBrokerMessageProto.getMessageStr();
        }
        else if (aBrokerMessageProto.hasMessageBin())
        {
            aBrokerMessage.Message = aBrokerMessageProto.getMessageBin().toByteArray();
        }
        
        return aBrokerMessage;
    }


    private byte[] serializeReliableMessage(ReliableMessage data)
    {
        ReliableMessageProto.Builder aBuilder = ReliableMessageProto.newBuilder()
                .setMessageId(data.MessageId)
                .setMessageType(data.MessageType.toString());
        if (data.Message != null)
        {
            if (data.Message instanceof String)
            {
                aBuilder.setMessageStr((String)data.Message);
            }
            else
            {
                aBuilder.setMessageBin(ByteString.copyFrom((byte[])data.Message));
            }
        }
        ReliableMessageProto aReliableMessageProto = aBuilder.build();
        
        return aReliableMessageProto.toByteArray();
    }

    private ReliableMessage deserializeReliableMessage(byte[] data) throws InvalidProtocolBufferException
    {
        ReliableMessageProto aReliableMessageProto = ReliableMessageProto.parseFrom(data);
        ReliableMessage aReliableMessage = new ReliableMessage();
        aReliableMessage.MessageId = aReliableMessageProto.getMessageId();
        aReliableMessage.MessageType = ReliableMessage.EMessageType.valueOf(aReliableMessageProto.getMessageType());
        if (aReliableMessageProto.hasMessageStr())
        {
            aReliableMessage.Message = aReliableMessageProto.getMessageStr();
        }
        else if (aReliableMessageProto.hasMessageBin())
        {
            aReliableMessage.Message = aReliableMessageProto.getMessageBin().toByteArray();
        }
        
        return aReliableMessage;
    }


    private byte[] serializeMonitorChannelMessage(MonitorChannelMessage data)
    {
        MonitorChannelMessageProto.Builder aBuilder = MonitorChannelMessageProto.newBuilder()
                .setMessageType(data.MessageType.toString());
        if (data.MessageContent != null)
        {
            if (data.MessageContent instanceof String)
            {
                aBuilder.setMessageContentStr((String)data.MessageContent);
            }
            else
            {
                aBuilder.setMessageContentBin(ByteString.copyFrom((byte[])data.MessageContent));
            }
        }
        MonitorChannelMessageProto aMonitorChannelMessageProto = aBuilder.build();
        
        return aMonitorChannelMessageProto.toByteArray();
    }

    private MonitorChannelMessage deserializeMonitorChannelMessage(byte[] data) throws InvalidProtocolBufferException
    {
        MonitorChannelMessageProto aMonitorChannelMessageProto = MonitorChannelMessageProto.parseFrom(data); 
        MonitorChannelMessage aMonitorChannelMessage = new MonitorChannelMessage();
        aMonitorChannelMessage.MessageType = MonitorChannelMessageType.valueOf(aMonitorChannelMessageProto.getMessageType());
        if (aMonitorChannelMessageProto.hasMessageContentStr())
        {
            aMonitorChannelMessage.MessageContent = aMonitorChannelMessageProto.getMessageContentStr();
        }
        else if (aMonitorChannelMessageProto.hasMessageContentBin())
        {
            aMonitorChannelMessage.MessageContent = aMonitorChannelMessageProto.getMessageContentBin().toByteArray();
        }
        
        return aMonitorChannelMessage;
    }
    
    
    private byte[] serializeVoidMessage(VoidMessage data)
    {
        VoidMessageProto aVoidMessageProto = VoidMessageProto.newBuilder()
                .build();
        return aVoidMessageProto.toByteArray();
    }
    
    private VoidMessage deserializeVoidMessage(byte[] data) throws InvalidProtocolBufferException
    {
        VoidMessageProto.parseFrom(data);
        VoidMessage aVoidMessage = new VoidMessage();
        return aVoidMessage;
    }
}

File: tags/EneterProtoBuf 6.0.0/EneterProtoBufSerializer/EneterProtoBufSerializer/ProtoBuf/ProtoBufSerializer.cs
﻿/*
 * Project: Eneter.ProtoBuf.Serializer
 * Author:  Ondrej Uzovic
 * 
 * Copyright © Ondrej Uzovic 2013
*/

using System;
using System.IO;
using Eneter.Messaging.DataProcessing.Serializing;
using Eneter.Messaging.EndPoints.Rpc;
using Eneter.Messaging.EndPoints.TypedMessages;
using Eneter.Messaging.MessagingSystems.Composites.MonitoredMessagingComposit;
using Eneter.Messaging.Nodes.Broker;
using Eneter.Messaging.Nodes.ChannelWrapper;
using ProtoBuf;

namespace Eneter.ProtoBuf
{
    /// <summary>
    /// Implements protocol buffer serialization for Eneter.Messaging.Framework.
    /// </summary>
    public class ProtoBufSerializer : ISerializer
    {
        /// <summary>
        /// Serializes data using Protocol Buffer binary format.
        /// </summary>
        /// <typeparam name="_T">data type of serialized data</typeparam>
        /// <param name="dataToSerialize">protocol buffer generated data class to be serialized.</param>
        /// <returns>array containing serialized data</returns>
        public object Serialize<_T>(_T dataToSerialize)
        {
            object aSerializedData;

            // If it is an internal Eneter type adapt it for ProtoBuf
            if (dataToSerialize is RpcMessage)
            {
                aSerializedData = SerializeRpcMessage(dataToSerialize as RpcMessage);
            }
            else if (dataToSerialize is EventArgs)
            {
                aSerializedData = SerializeEventArgs(dataToSerialize as EventArgs);
            }
            else if (dataToSerialize is WrappedData)
            {
                aSerializedData = SerializeWrappedData(dataToSerialize as WrappedData);
            }
            else if (dataToSerialize is BrokerMessage)
            {
                aSerializedData = SerializeBrokerMessage(dataToSerialize as BrokerMessage);
            }
            else if (dataToSerialize is ReliableMessage)
            {
                aSerializedData = SerializeReliableMessage(dataToSerialize as ReliableMessage);
            }
            else if (dataToSerialize is MonitorChannelMessage)
            {
                aSerializedData = SerializeMonitorChannelMessage(dataToSerialize as MonitorChannelMessage);
            }
            else if (dataToSerialize is VoidMessage)
            {
                aSerializedData = SerializeVoidMessage(dataToSerialize as VoidMessage);
            }
            else
            {
                // If it is not internal Eneter message use directly ProtoBuf serializer.
                aSerializedData = SerializeProtoBuf<_T>(dataToSerialize);
            }

            return aSerializedData;
        }

        /// <summary>
        /// Deserializes data from the Protocol Buffer binary format.
        /// </summary>
        /// <typeparam name="_T">data type of deserialized data.</typeparam>
        /// <param name="serializedData">data (byte[]) serialized by protocol buffer to be deserialized</param>
        /// <returns>instance of deserialized data type</returns>
        public _T Deserialize<_T>(object serializedData)
        {
            using (MemoryStream aBuf = new MemoryStream((byte[])serializedData))
            {
                _T aDeserializedObject;

                // If it is an internal Eneter type, adapt it from the proto type.
                if (typeof(_T) == typeof(RpcMessage))
                {
                    aDeserializedObject = (_T)((object)DeserializeRpcMessage(aBuf));
                }
                else if (typeof(_T) == typeof(EventArgs))
                {
                    aDeserializedObject = (_T)((object)DeserializeEventArgs(aBuf));
                }
                else if (typeof(_T) == typeof(WrappedData))
                {
                    aDeserializedObject = (_T)((object)DeserializeWrappedData(aBuf));
                }
                else if (typeof(_T) == typeof(BrokerMessage))
                {
                    aDeserializedObject = (_T)((object)DeserializeBrokerMessage(aBuf));
                }
                else if (typeof(_T) == typeof(ReliableMessage))
                {
                    aDeserializedObject = (_T)((object)DeserializeReliableMessage(aBuf));
                }
                else if (typeof(_T) == typeof(MonitorChannelMessage))
                {
                    aDeserializedObject = (_T)((object)DeserializeMonitorChannelMessage(aBuf));
                }
                else if (typeof(_T) == typeof(VoidMessage))
                {
                    aDeserializedObject = (_T)((object)DeserializeVoidMessage(aBuf));
                }
                else
                {
                    // If it is not internal Eneter message use directly ProtoBuf serializer.
                    aDeserializedObject = Serializer.Deserialize<_T>(aBuf);
                }

                return aDeserializedObject;
            }
        }

        private byte[] SerializeRpcMessage(RpcMessage data)
        {
            RpcMessageProto anRpcMessageProto = new RpcMessageProto();
            anRpcMessageProto.Id = data.Id;
            anRpcMessageProto.Flag = data.Flag;
            anRpcMessageProto.OperationName = data.OperationName;
            anRpcMessageProto.Error = data.Error;
            if (data.SerializedData != null)
            {
                foreach (byte[] aMethodParameter in data.SerializedData)
                {
                    anRpcMessageProto.SerializedData.Add(aMethodParameter);
                }
            }

            return SerializeProtoBuf<RpcMessageProto>(anRpcMessageProto);
        }

        private RpcMessage DeserializeRpcMessage(MemoryStream data)
        {
            RpcMessageProto anRpcMessageProto = Serializer.Deserialize<RpcMessageProto>(data);
            RpcMessage anRpcMessage = new RpcMessage();
            anRpcMessage.Id = anRpcMessageProto.Id;
            anRpcMessage.Flag = anRpcMessageProto.Flag;
            anRpcMessage.OperationName = anRpcMessageProto.OperationName;
            anRpcMessage.Error = anRpcMessageProto.Error;
            
            if (anRpcMessageProto.SerializedData != null)
            {
                anRpcMessage.SerializedData = new object[anRpcMessageProto.SerializedData.Count];
                for (int i = 0; i < anRpcMessageProto.SerializedData.Count; ++i)
                {
                    anRpcMessage.SerializedData[i] = anRpcMessageProto.SerializedData[i];
                }
            }

            return anRpcMessage;
        }

        private byte[] SerializeEventArgs(EventArgs data)
        {
            return myEmptyMessage;
        }

        private EventArgs DeserializeEventArgs(MemoryStream data)
        {
            Serializer.Deserialize<EventArgsProto>(data);
            return myEventArgs;
        }

        private byte[] SerializeWrappedData(WrappedData data)
        {
            WrappedDataProto aWrappedDataProto = new WrappedDataProto();
            aWrappedDataProto.AddedData = (string)data.AddedData;
            if (data.OriginalData != null)
            {
                if (data.OriginalData is string)
                {
                    aWrappedDataProto.OriginalDataStr = (string)data.OriginalData;
                }
                else
                {
                    aWrappedDataProto.OriginalDataBin = (byte[])data.OriginalData;
                }
            }
            return SerializeProtoBuf<WrappedDataProto>(aWrappedDataProto);
        }

        private WrappedData DeserializeWrappedData(MemoryStream data)
        {
            WrappedDataProto aWrappedDataProto = Serializer.Deserialize<WrappedDataProto>(data);
            WrappedData aWrappedData = new WrappedData();
            aWrappedData.AddedData = aWrappedDataProto.AddedData;
            if (aWrappedDataProto.OriginalDataStrSpecified)
            {
                aWrappedData.OriginalData = aWrappedDataProto.OriginalDataStr;
            }
            else if (aWrappedDataProto.OriginalDataBinSpecified)
            {
                aWrappedData.OriginalData = aWrappedDataProto.OriginalDataBin;
            }

            return aWrappedData;
        }


        private byte[] SerializeBrokerMessage(BrokerMessage data)
        {
            BrokerMessageProto aBrokerMessageProto = new BrokerMessageProto();
            aBrokerMessageProto.Request = data.Request.ToString();
            aBrokerMessageProto.MessageTypes.AddRange(data.MessageTypes);
            if (data.Message is string)
            {
                aBrokerMessageProto.MessageStr = (string)data.Message;
            }
            else
            {
                aBrokerMessageProto.MessageBin = (byte[])data.Message;
            }
            return SerializeProtoBuf<BrokerMessageProto>(aBrokerMessageProto);
        }

        private BrokerMessage DeserializeBrokerMessage(MemoryStream data)
        {
            BrokerMessageProto aBrokerMessageProto = Serializer.Deserialize<BrokerMessageProto>(data);
            BrokerMessage aBrokerMessage = new BrokerMessage();
            aBrokerMessage.Request = (EBrokerRequest) Enum.Parse(typeof(EBrokerRequest), aBrokerMessageProto.Request, true);
            aBrokerMessage.MessageTypes = aBrokerMessageProto.MessageTypes.ToArray();
            if (aBrokerMessageProto.MessageStrSpecified)
            {
                aBrokerMessage.Message = aBrokerMessageProto.MessageStr;
            }
            else if (aBrokerMessageProto.MessageBinSpecified)
            {
                aBrokerMessage.Message = aBrokerMessageProto.MessageBin;
            }
            return aBrokerMessage;
        }


        private byte[] SerializeReliableMessage(ReliableMessage data)
        {
            ReliableMessageProto aReliableMessageProto = new ReliableMessageProto();
            aReliableMessageProto.MessageId = data.MessageId;
            aReliableMessageProto.MessageType = data.MessageType.ToString();
            if (data.Message is string)
            {
                aReliableMessageProto.MessageStr = (string)data.Message;
            }
            else
            {
                aReliableMessageProto.MessageBin = (byte[])data.Message;
            }
            return SerializeProtoBuf<ReliableMessageProto>(aReliableMessageProto);
        }

        private ReliableMessage DeserializeReliableMessage(MemoryStream data)
        {
            ReliableMessageProto aReliableMessageProto = Serializer.Deserialize<ReliableMessageProto>(data);
            ReliableMessage aReliableMessage = new ReliableMessage();
            aReliableMessage.MessageId = aReliableMessageProto.MessageId;
            aReliableMessage.MessageType = (ReliableMessage.EMessageType)Enum.Parse(typeof(ReliableMessage.EMessageType), aReliableMessageProto.MessageType, true);
            if (aReliableMessageProto.MessageStrSpecified)
            {
                aReliableMessage.Message = aReliableMessageProto.MessageStr;
            }
            else if (aReliableMessageProto.MessageBinSpecified)
            {
                aReliableMessage.Message = aReliableMessageProto.MessageBin;
            }
            return aReliableMessage;
        }


        private byte[] SerializeMonitorChannelMessage(MonitorChannelMessage data)
        {
            MonitorChannelMessageProto aMonitorChannelMessageProto = new MonitorChannelMessageProto();
            aMonitorChannelMessageProto.MessageType = data.MessageType.ToString();
            if (data.MessageContent is string)
            {
                aMonitorChannelMessageProto.MessageContentStr = (string)data.MessageContent;
            }
            else
            {
                aMonitorChannelMessageProto.MessageContentBin = (byte[])data.MessageContent;
            }
            return SerializeProtoBuf<MonitorChannelMessageProto>(aMonitorChannelMessageProto);
        }

        private MonitorChannelMessage DeserializeMonitorChannelMessage(MemoryStream data)
        {
            MonitorChannelMessageProto aMonitorChannelMessageProto = Serializer.Deserialize<MonitorChannelMessageProto>(data);
            MonitorChannelMessage aMonitorChannelMessage = new MonitorChannelMessage();
            aMonitorChannelMessage.MessageType = (MonitorChannelMessageType)Enum.Parse(typeof(MonitorChannelMessageType), aMonitorChannelMessageProto.MessageType, true);
            if (aMonitorChannelMessageProto.MessageContentStrSpecified)
            {
                aMonitorChannelMessage.MessageContent = aMonitorChannelMessageProto.MessageContentStr;
            }
            else if (aMonitorChannelMessageProto.MessageContentBinSpecified)
            {
                aMonitorChannelMessage.MessageContent = aMonitorChannelMessageProto.MessageContentBin;
            }
            return aMonitorChannelMessage;
        }

        private byte[] SerializeVoidMessage(VoidMessage data)
        {
            return myEmptyMessage;
        }

        private VoidMessage DeserializeVoidMessage(MemoryStream data)
        {
            // Read bytes from the stream.
            Serializer.Deserialize<VoidMessageProto>(data);
            return myVoidMessage;
        }

        private byte[] SerializeProtoBuf<_T>(_T dataToSerialize)
        {
            using (MemoryStream aBuf = new MemoryStream())
            {
                Serializer.Serialize<_T>(aBuf, dataToSerialize);
                return aBuf.ToArray();
            }
        }

        private readonly EventArgs myEventArgs = new EventArgs();
        private readonly VoidMessage myVoidMessage = new VoidMessage();
        private readonly byte[] myEmptyMessage = { 8, 0 };
    }
}
